### ISSUE TITLE
Calling procedures

### ISSUE THREAD

**Issue by gbdlin** [bitfocus/companion#3149]
### Is this a feature relevant to companion itself, and not a module?

- [x] I believe this to be a feature for companion, not a module

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the feature

Currently the only way to reuse sets of actions is to call one button from another button. As this approach works pretty well, it has some major drawbacks: it can be easily broken (by rearranging pages or buttons that rely on each other), it is hard to trace and doesn't allow parametrized inputs.

Introducing procedures that are just set of actions that will be triggered one by one just like on buttons would fix those issues. Each procedure should accept a configurable amount of parameters that would be passed from button, trigger or another procedure and those parameters would be present as variables in the procedure itself. 

It is worth noting that implementation should account for 2 instances of the same procedure being called at once and make sure parameters are locally scope and each call sees their own values.

This issue can possibly benefit people waiting for #3138, #3085, #2370, #2138 (by exposing procedures over HTTP API), #1924, #1676 and can be a step towards #2387. Also loosely relates to #1909 as the concept of passing parameters seems similar. It may also help with #1849. 

### Usecases

Example: with ATEM switchers user may want to set a specific output as a flying key on a specific position on the screen and then trigger a specific transition, maybe changing also the background. Now, if there are multiple inputs that need this treatment, the only way is to separate the same actions onto yet another button (possibly hidden on an unused page) and trigger this button from other buttons and have only the actions that differ on each directly pressed button.

Procedures will allow to streamline this process as well as parametrize the input, so now only action on buttons would be triggering a specific procedure with a different parameter. Changing the position of the flying key would also be easier as it can be another parameter on the same procedure.

**Comment by tylernap** [comment]
Was flipping through issues here as I was thinking of something like this but didn't know how to word it. I think this is pretty close. From a programmer's perspective, it would essentially be a method that can be reused for various buttons.

I was preparing for an event making buttons and ended up with some pretty complex logic to make it as idiot-proof as I could (untrained volunteers were expected to use it (not to say they are idiots)) and I had to make sheets of buttons for various surfaces. Every time I realized I had to make a change to the logic, I felt a bit of my soul leave my body as I knew I had to go through every button individually and unwrap the nested logic and action groups to make the changes I needed. Having a procedure or method that I could update in one place would have saved me so much time. Also, given that local variables have been released in v4.1, this sounds like a good time to revisit something like this as parameter inputs could be treated as local variables.

This could probably be ignored for an initial release, but at least in my use case, being able to pass a connection as a parameter as well would be useful. I'm sure there are other implications with that though and I am not even sure how possible that would be.

Anyway, +1 to this

**Comment by AshtonBridgeman** [comment]
+1
We have a very complicated companion setup with lots of duplicated logic and buttons pressing buttons, all to try and achieve a sense of re-usable functions (or procedures as you are calling them). This is exceptionally hard to follow when debugging or working on existing automations. I feel a node graph view would be well suited to something like this
